### ISSUE TITLE
Please remove dependencies on autoloaders

### ISSUE THREAD

**Issue by cubiclesoft** [phpseclib/phpseclib#973]
phpseclib has gotten a bit too dependency-heavy as of late.  Currently, to instantiate the RSA class without an autoloader and/or Composer, a user has to do something along the lines of:

```php
	$rootpath = str_replace("\\", "/", dirname(__FILE__));

	require_once $rootpath . "/path/to/phpseclib/Crypt/RSA.php";
	require_once $rootpath . "/path/to/phpseclib/Crypt/Hash.php";

	$path = $rootpath . "/path/to/phpseclib/Crypt/RSA/";
	$dir = opendir($path);
	if ($dir)
	{
		while (($file = readdir($dir)) !== false)
		{
			if (substr($file, -4) === ".php")  require_once $path . $file;
		}

		closedir($dir);
	}

	require_once $rootpath . "/path/to/phpseclib/Math/BigInteger.php";

	$rsa = new phpseclib\Crypt\RSA();
	...
```

To have any hope of future compatibility, the opendir()/readdir()/closedir() bit is necessary.  The files in the RSA directory might as well be bundled with the RSA class since instantiating/using RSA() throws fits until every last file in the RSA directory is loaded.

A possible solution from within your classes could look like:

```php
	if (!class_exists("phpseclib\\Crypt\\RSA\\MSBLOB", false))  require_once str_replace("\\", "/", dirname(__FILE__)) . "/RSA/MSBLOB.php";
```

OR:

```php
	if (!class_exists("phpseclib\\Crypt\\RSA\\MSBLOB", false) && file_exists(str_replace("\\", "/", dirname(__FILE__)) . "/RSA/MSBLOB.php"))  require_once str_replace("\\", "/", dirname(__FILE__)) . "/RSA/MSBLOB.php";
```

You probably know the most common use-cases for each class in phpseclib and therefore can preload the appropriate classes as-needed with something like the above without resorting to an autoloader.  Inclusion might even be as early as __construct() (i.e. long before an autoloader would ever be triggered).  Your library also knows where all of its files are supposed to be in relation to each other.

Not everyone wants to use autoloaders, Composer, PSRs, etc.  Some of us actually prefer small, compact libraries and don't mind a few 'require_once's here and there.

**Comment by terrafrost** [comment]
You're talking about the master branch. The master branch isn't going to be a solution for everyone. The 1.0 branch doesn't require autoloaders nor does it even really require PHP5. In-so-far as any branch is a solution for everyone it'd be the 1.0 branch.

Are there specific features from the master branch that you're wanting that the 1.0 branch doesn't have?

> The files in the RSA directory might as well be bundled with the RSA class since instantiating/using RSA() throws fits until every last file in the RSA directory is loaded.

If you're just manually copy / pasting the files into a directory you could just delete the files you don't need from the RSA subdirectory. Here's how it knows they exist (from the `_initialize_static_variables` method):

```php
                foreach (glob(__DIR__ . '/RSA/*.php') as $file) {
                    $name = pathinfo($file, PATHINFO_FILENAME);
                    $type = 'phpseclib\Crypt\RSA\\' . $name;
                    $meta = new \ReflectionClass($type);
```
So if you delete a file it won't be returned with `glob`. (`ReflectionClass` is what is triggering the auto-loader).

**Comment by cubiclesoft** [comment]
There's no particular feature I want/need.  In general, I use the master branch of a repo and wait for a relatively stable variant of it to show up.  I do that mostly because master generally has better support options - that is, generally-speaking, bugs get fixed and the developer cares.  It's hard to tell with most projects where support is at with older versions because it is rare to see a deprecation timeline and even rarer to see someone supporting multiple versions and backporting code.  Most people don't have the time nor the desire to develop/manage/maintain multiple versions of a project.

glob()'ing the RSA directory and calling ReflectionClass tosses a monkey wrench into any script that might turn the library into a 'standalone' variant (i.e. merging multiple PHP files into one file) because the RSA directory would cease to exist but the classes from the files in that directory would be generally available.  I prefer standalone builds when I can get them (rarer these days).

Standalone build issue aside, how about a simple 'class_exists'/'require_once' line right before the ReflectionClass() usage?  You already know exactly where the file is located on disk since the code just glob()'ed the whole directory.

On a loosely related note, while looking at the code and locating other ReflectionClass references, I noticed that RSA::addFileFormat() appears to be broken:

```php
    static function addFileFormat($fullname)
    {
        self::_initialize_static_variables();

        if (class_exists($fullname)) {
            $meta = new \ReflectionClass($path);
            ...
```

$path is not defined before being used.

**Comment by bantu** [comment]
Including all files is most likely a waste of time and memory. Manually managing dependencies is non-trivial and has resulted in countless problems for this very project. Besides that, it is just unneccesarry code that can be removed by introducing proper abstraction. This abstraction is exactly what autoloading provides. You are free to shoot yourself in the foot by not using an autoloader, but you're on your own then. I don't even understand why @terrafrost is trying to make mediocre workarounds to autoloading work for you.

The master branch is the development branch. If you want something stable, use the 2.0 branch. As far as I know the composer autoloader (https://github.com/composer/composer/blob/master/src/Composer/Autoload/ClassLoader.php) works standalone without the rest of composer. Properly documenting how to use it is an open task in #257, feel free to tackle it.

**Comment by cubiclesoft** [comment]
> Including all files is most likely a waste of time and memory.

```
PHP Fatal error:  Uncaught exception 'ReflectionException' with message 'Class phpseclib\Crypt\RSA\MSBLOB does not exist' in D:\WEB\test\support\phpseclib\Crypt\RSA.php:360
Stack trace:
#0 D:\WEB\test\support\phpseclib\Crypt\RSA.php(360): ReflectionClass->__construct('phpseclib\\Crypt...')
#1 D:\WEB\test\support\phpseclib\Crypt\RSA.php(382): phpseclib\Crypt\RSA::_initialize_static_variables()
#2 D:\WEB\test\install.php(30): phpseclib\Crypt\RSA->__construct()
#3 {main}
  thrown in D:\WEB\test\support\phpseclib\Crypt\RSA.php on line 360
```

The RSA class already attempts to load ALL of the files in the RSA directory during object construction.  I'm not saying to load *everything* across all of phpseclib which would certainly be wasteful.  I'm saying to eliminate the obvious in specific areas where the code is going to definitely load a lot of *something*.  How is the solution I proposed a problem?  After all, we're talking about a couple of carefully placed lines of code with extreme knowledge of the file structure - not some vague topic related to autoloaders.

**Comment by terrafrost** [comment]
> It's hard to tell with most projects where support is at with older versions because it is rare to see a deprecation timeline and even rarer to see someone supporting multiple versions and backporting code.

Feel free to look at the commit history. The 1.0 branch is being actively maintained. In fact, 21 days ago I did backport some code from a PR request made against the master branch that I had merged:

https://github.com/phpseclib/phpseclib/commit/ac97ce5d69888f785b1db37d1c4d4dc34e942aff

Normally I'd have cherry picked the commit but I didn't realize the code was needed in the 1.0 branch until after I had pushed to master.

> $path is not defined before being used.

Nice find. I'll look into that as time permits.

> The RSA class already attempts to load ALL of the files in the RSA directory during object construction. I'm not saying to load everything across all of phpseclib which would certainly be wasteful. I'm saying to eliminate the obvious in specific areas where the code is going to definitely load a lot of something.

Well if you have a *better* approach do feel free to suggest it. The approach you've suggested, however, is distinctly /not/ better, however, in-so-far as I'm concerned.

> How is the solution I proposed a problem?

Because it kinda neuters phpseclib's ability to effectively use the whole composer / packagist ecosystem. In the master and 2.0 branches I added paragonie/random_compat as a required library because, frankly, I don't want to spend a ton of time trying to make the best most cryptographically secure RNG ever. phpseclib is a fairly expansive project and there's a lot to manage already and in some cases I may be willing to offload components that I simply don't care to maintain to other composer packages. I won't do that for the 1.0 branch and for the 2.0 branch I'm thinking I'll only do that for PSR-4 autoloadable stuff (composer does more than just PSR-4) but for master / 3.0...  composer all the way.

Sure, I could download the random_compat repo and add it to phpseclib myself but then what if random_compat has an update? Do I do that every time random_compat has an update?  composer kinda solves these problems.
